### PR TITLE
openid audience claim check needs to support []interface{} type.

### PR DESCRIPTION
**Pull request by jmazzitelli**
json unmarshalling produces []interface{}, we need to support that type

fixes: https://github.com/kiali/kiali/issues/6226

### CODE REVIEW

**Comment by jmazzitelli** [comment]
new unit tests pass... also ran auth-related molecule tests to confirm nothing broke

```
=====================
=== TEST RESULTS: ===
=====================

                             openid-test... success [1m 50s]
                              token-test... success [2m 12s]
                        header-auth-test... success [1m 54s]

Test logs can be found at: /tmp/kiali-molecule-test-logs.2023-06-01_15-07-34

```

File: business/authentication/openid_auth_controller.go [hunk 1]
```diff
@@ -1236,6 +1236,38 @@ func parseTimeClaim(claimValue interface{}) (int64, error) {
...
+				return fmt.Errorf("the OpenId string token is not targeted for Kiali; got []aud = '%v'", audienceClaim)
+			}
+		case []any:
+			if len(audienceClaim.([]any)) != 1 {
+				return fmt.Errorf("the OpenId token was rejected because it has more than one audience; got aud = %v", audienceClaim)
+			}
+			acStr := fmt.Sprintf("%v", audienceClaim.([]any)[0])
+			if oidCfg.ClientId != acStr {
+				return fmt.Errorf("the OpenId token is not targeted for Kiali; got []aud = '%v'", acStr)
+			}
```
Comment (jmazzitelli) [inline business/authentication/openid_auth_controller.go:1262]: This is essentially the code that was added to fix the issue.

**Review by nrfox — COMMENTED** [review]
One small nit otherwise LGTM

**Comment by jmazzitelli** [comment]
FWIW, here is the test code I tried to get working. It is almost there, but failed with an error that I can't figure out how to workaround in the mock request. I give up. Others are free to try to complete this test. I was going to have it submit one good request and then one bad one (the bad one with an invalid "aud" claim). That's where there is the unused/commented out "openIdTestTokenWithBadAud". I never got far enough to be able to use that.

```go
func TestValidateOpenIdTokenInHouse(t *testing.T) {
	// THIS TEST FAILS WITH "the OpenId token is missing the kid header claim" !!!

	// These tokens were generated via https://dinochiesa.github.io/jwt/ using private/public keys generated via https://mkjwk.org/
	//
	// {
	//  "sub": "jdoe@domain.com",
	//  "name": "John Doe",
	//  "iat": 1516239022,
	//  "nonce": "1ba9b834d08ac81feb34e208402eb18e909be084518c328510940184",
	//  "exp": 1638316801,
	//  "iss": "http://127.0.0.1:33333",
	//  "aud": "kiali-client"
	// }
	openIdTestTokenWithGoodAud := "eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJqZG9lQGRvbWFpbi5jb20iLCJuYW1lIjoiSm9obiBEb2UiLCJpYXQiOjE1MTYyMzkwMjIsIm5vbmNlIjoiMWJhOWI4MzRkMDhhYzgxZmViMzRlMjA4NDAyZWIxOGU5MDliZTA4NDUxOGMzMjg1MTA5NDAxODQiLCJleHAiOjE2MzgzMTY4MDEsImlzcyI6Imh0dHA6Ly8xMjcuMC4wLjE6MzMzMzMiLCJhdWQiOiJraWFsaS1jbGllbnQifQ.QBARZtXEOYZGHv02gQWYcUEOooFaOSKnCozrIQCXmhufvuI5yOmHqIy9BcpqeGyVtlCr8SkA1kW900bIfqvnMg"

	// {
	//  "sub": "jdoe@domain.com",
	//  "name": "John Doe",
	//  "iat": 1516239022,
	//  "nonce": "1ba9b834d08ac81feb34e208402eb18e909be084518c328510940184",
	//  "exp": 1638316801,
	//  "iss": "http://127.0.0.1:33333",
	//  "aud": "bad-aud-client"
	// }
	//openIdTestTokenWithBadAud := "eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJqZG9lQGRvbWFpbi5jb20iLCJuYW1lIjoiSm9obiBEb2UiLCJpYXQiOjE1MTYyMzkwMjIsIm5vbmNlIjoiMWJhOWI4MzRkMDhhYzgxZmViMzRlMjA4NDAyZWIxOGU5MDliZTA4NDUxOGMzMjg1MTA5NDAxODQiLCJleHAiOjE2MzgzMTY4MDEsImlzcyI6Imh0dHA6Ly8xMjcuMC4wLjE6MzMzMzMiLCJhdWQiOiJiYWQtYXVkLWNsaWVudCJ9.ea4FW7Dsuwt2M_cRRcFW_NPpQ7pBuSdvvGrml3Zp8vnsQ1qLMtr09MfXPaHc1x29HtkFmB9g8TMeXlO0LJ1EZQ"

	// we start with a good token - our second test will switch this to the token with the bad aud value
	openIdTestTokenToUse := openIdTestTokenWithGoodAud

	cachedOpenIdMetadata = nil
	var oidcMetadata []byte
	testServer := httptest.NewUnstartedServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path == "/.well-known/openid-configuration" {
			w.WriteHeader(200)
			_, _ = w.Write(oidcMetadata)
		}
		if r.URL.Path == "/token" {
			_ = r.ParseForm()
			assert.Equal(t, "f0code", r.Form.Get("code"))
			assert.Equal(t, "authorization_code", r.Form.Get("grant_type"))
			assert.Equal(t, "kiali-client", r.Form.Get("client_id"))
			assert.Equal(t, "https://kiali.io:44/kiali-test", r.Form.Get("redirect_uri"))

			w.WriteHeader(200)
			_, _ = w.Write([]byte("{ \"id_token\": \"" + openIdTestTokenToUse + "\" }"))
		}
	}))
	defer testServer.Close()

	// because we have a hardcoded token for this test that is pre-encrypted with the issuer URL, we need to start the server on that same URL
	testServerListenerFixedPort, err := net.Listen("tcp", "127.0.0.1:33333")
	assert.Nil(t, err, "Cannot start test server on fixed port")
	testServer.Listener = testServerListenerFixedPort
	testServer.Start()

	oidcMeta := openIdMetadata{
		Issuer:                 testServer.URL,
		AuthURL:                testServer.URL + "/auth",
		TokenURL:               testServer.URL + "/token",
		JWKSURL:                testServer.URL + "/jwks",
		UserInfoURL:            "",
		Algorithms:             nil,
		ScopesSupported:        []string{"openid"},
		ResponseTypesSupported: []string{"code"},
	}
	oidcMetadata, err = json.Marshal(oidcMeta)
	assert.Nil(t, err)

	clockTime := time.Date(2021, 12, 1, 0, 0, 0, 0, time.UTC)
	util.Clock = util.ClockMock{Time: clockTime}

	conf := config.NewConfig()
	conf.Server.WebRoot = "/kiali-test"
	conf.LoginToken.SigningKey = "kiali67890123456"
	conf.LoginToken.ExpirationSeconds = 1
	conf.Auth.OpenId.IssuerUri = testServer.URL
	conf.Auth.OpenId.ClientId = "kiali-client"
	conf.Auth.OpenId.DisableRBAC = true // true is needed to trigger the call to validateOpenIdTokenInHouse
	config.Set(conf)

	// Returning some namespace when a cluster API call is made should have the result of
	// a successful authentication.
	k8s := kubetest.NewK8SClientMock()
	mockClientFactory := kubetest.NewK8SClientFactoryMock(k8s)
	business.SetWithBackends(mockClientFactory, nil)

	k8s.On("GetProjects", "").Return([]osproject_v1.Project{
		{ObjectMeta: meta_v1.ObjectMeta{Name: "Foo"}},
	}, nil)

	stateHash := sha256.Sum224([]byte(fmt.Sprintf("%s+%s+%s", "nonceString", clockTime.UTC().Format("060102150405"), config.GetSigningKey())))
	uri := fmt.Sprintf("https://kiali.io:44/api/authenticate?code=f0code&state=%x-%s", stateHash, clockTime.UTC().Format("060102150405"))
	request := httptest.NewRequest(http.MethodGet, uri, nil)
	request.AddCookie(&http.Cookie{
		Name:  OpenIdNonceCookieName,
		Value: "nonceString",
	})

	controller := NewOpenIdAuthController(CookieSessionPersistor{}, func(authInfo *api.AuthInfo) (*business.Layer, error) {
		if authInfo.Token != openIdTestTokenToUse {
			return nil, errors.New("unexpected token")
		}
		k8sclients := make(map[string]kubernetes.ClientInterface)
		k8sclients[conf.KubernetesConfig.ClusterName] = k8s
		return business.NewWithBackends(k8sclients, k8sclients, nil, nil), nil
	})

	rr := httptest.NewRecorder()
	controller.GetAuthCallbackHandler(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		assert.Failf(t, "Callback function shouldn't have been called.", "")
	})).ServeHTTP(rr, request)

	expectedExpiration := time.Date(2021, 12, 1, 0, 0, 1, 0, time.UTC)

	// Check that cookies are set and have the right expiration.
	response := rr.Result()
	assert.Len(t, response.Cookies(), 2)

	// nonce cookie cleanup
	assert.Equal(t, OpenIdNonceCookieName, response.Cookies()[0].Name)
	assert.True(t, clockTime.After(response.Cookies()[0].Expires))

	// Session cookie
	assert.Equal(t, AESSessionCookieName, response.Cookies()[1].Name)
	assert.Equal(t, expectedExpiration, response.Cookies()[1].Expires)
	assert.Equal(t, http.StatusFound, response.StatusCode)

	// Redirection to boot the UI
	assert.Equal(t, "/kiali-test/", response.Header.Get("Location"))
}
```

File: business/authentication/openid_auth_controller.go [hunk 2]
```diff
@@ -1236,6 +1236,38 @@ func parseTimeClaim(claimValue interface{}) (int64, error) {
...
+		default:
+			return fmt.Errorf("the OpenId token has an unexpected audience claim; got '%v'", audienceClaim)
```
[the remark this answers is not in the document]
Comment (jmazzitelli) [reply business/authentication/openid_auth_controller.go:1264]: yes, excellent point. I thought that before I even started fixing this... "I wish the type was logged here" :)

[the remark this answers is not in the document]
Comment (jmazzitelli) [reply business/authentication/openid_auth_controller.go]: done. type is logged with the value now:
```go
return fmt.Errorf("the OpenId token has an unexpected audience claim; value [%v] of type [%T]", audienceClaim, audienceClaim)
```

**Review by nrfox — APPROVED** [review]

**Comment by jmazzitelli** [comment]
@nrfox I was able to figure out how to get a unit test implemented that actually tests that audience claim check. New commit added to this PR - need another review/approval.

**Review by nrfox — APPROVED** [review]
Thanks for adding a test for this as well!

**Pull request merged**